Add OpenXR support to the core

BastiaanOlij opened the pull request:
This is a port of the OpenXR plugin to core, the main reason for this is the way OpenXR ties into Vulkan, with OpenXR wanting to take ownership of creating the Vulkan instance and managing the creation of swap chains for use in XR.
This requires OpenXR to be started and initialized at startup, for this reason new project settings have been added to ensure OpenXR is loaded. More settings will be added soon to further configure this. There is also a placeholder setting for running OpenXR in the editor, this for future plans to enable XR functionality within the editor.

The main OpenXR logic is implemented in the `drivers` folder through a new OpenXR device. This handles the OpenXR implementation itself.  Just like our original plugin the driver is implemented with extensions that implement specific parts of the implementation. There is for instance a Vulkan extension that implements the Vulkan bits of the solution. In due time we can create an OpenGL extension if the OpenGL renderer is used.

The Vulkan driver has been modified to use the Vulkan extension, when available, to set things up.

There is also an `OpenXRInterface` implementation in modules that implements the XRServer components.

I think this PR is ready for review and start doing the cleanup to get this to be merged. There are a few things left to do but I think these can have their own follow up PRs

- currently I'm doing a copy into the swapchain buffer instead of rendering directly into this. The main problem I have here is one of timing, OpenXR needs to have its session created and setup before we can create our swapchains. This is something that happens way too late as this will happen in the background, often waiting for the XR runtime to startup and switch over. It may require finishing #51179 and using our current approach to introduce the OpenXR swapchain just for the ARVR viewport. That said, OpenXR requires swapchains to be used for 2D UI viewports when using layers so there is more to do here.
- Action map and controller logic is finished but we need to add a UI to edit the actionmap
- Still issues with OpenXR wanting us to render in linear color space but linear color space on RGBA8 gives banding. 
- Should work on Windows and Linux but only tested SteamVR on Windows, but I've disabled Android and we can do this in a follow up PR. Note that currently we are assuming that we can use the Khronos loader on Android instead of using Metas proprietary loader for Quest.
- Port over features like hand tracking, pass through, composition layers, etc.

A demo project to test this work with can be found here: https://github.com/BastiaanOlij/godot4_openxr_demo

BastiaanOlij committed:
Add OpenXR 1.0.22 to thirdparty libraries

Will be compiled and used in the next commit.

Co-authored-by: Rémi Verschelde <rverschelde@gmail.com>

BastiaanOlij committed:
Implementing OpenXR driver

BastiaanOlij commented:
This is starting to work, it now is based on #51179, while there are some issues still around updating the framebuffers.

Two major problems at the moment:
- the timing is WAY off, it has a lot of problems doing proper predicting
- `keep_linear` has disappeared, now this might have to do with Vulkan being able to cater for this better, but OpenXR is expecting our data in linear color space, not converted to sRGB.

I have a number of things to polish up on 51179 first, but we're getting somewhere with this port.

m4gr3d commented:
@BastiaanOlij It may be worth splitting the `thirdpary/openxr/...` additions into a separate PR. I'm guessing it's mostly a dump, so no reviews should be needed for that section?
It'll make reviewing the actual integration easier.

BastiaanOlij commented:
> @BastiaanOlij It may be worth splitting the `thirdpary/openxr/...` additions into a separate PR. I'm guessing it's mostly a dump, so no reviews should be needed for that section? It'll make reviewing the actual integration easier.

It's in a separate commit, so if you just look at the last commit only, those are the real changes. 

First commit is a copy of (an older version of) 51179
Second commit is the openxr thirdparty library
Third commit is the real work.

m4gr3d commented:
Thanks for the clarification!

BastiaanOlij commented:
I've redone the last bit of this so it no longer used 51179 but does a copy of Godots results into the textures provided by OpenXR. This is slightly slower but it doesn't require us to do really ugly stuff.  Once I improve 51179 we'll revisit that. 

I'm still having the timing issues, I may try and move part of `process` closer to the moment we render to see if that solves it, but its very weird for sure.

Also I think Steam now doesn't like us providing the data in sRGB when its expecting linear color space so things are too bright. Still looking into that.

fire commented:
Noticed that changing vulkan vsync to mailbox mvoes the fps from 60fps to 90fps (HMD). Hope that helps.

BastiaanOlij commented:
> Noticed that changing vulkan vsync to mailbox moves the fps from 60fps to 90fps (HMD). Hope that helps.

Yeah we need to find out how to turn vsync off all together. OpenXR will set the pace.

fire commented:
Sent you the branch for tracker velocity including the head.

I think the bug is the engine tick is not the tracker (head) tick from the openxr runtime.

fire commented:
Discussed with @BastiaanOlij. CC @lyuma 

To get CICD to pass, it's not as simple as disabling the platforms because the OpenXR driver is tied into the renderer.

The plan is to swap every platform other than Linux and Windows with the dummy OpenXR.

Android needs a lot more work, and we need help there. @m4gr3d 

We need to create a dummy implementation for the disabled platforms. There is a dummy implementation, but it doesn't compile because it's far from complete. Currently, the dummy OpenXR implementation re-uses the header files.

The right solution is to create a dummy folder and put header files with just the class definitions and the few methods in the app but without code. Scons can then be instructed to use the dummy folder instead of the main folder.

The platforms without OpenXR support will then compile fine.

BastiaanOlij commented:
Using a separate folder for the dummy driver won't allow us to change the header files afterall because of full path names being used when including, opted for an #ifdef approach.

Still no closer to figuring out why the timing is not working properly

BastiaanOlij commented:
Grmbl, whats wrong with the macro that the compiler on Linux hates it?

akien-mga commented on modules/openxr/openxr_action_sets.cpp:
```diff
@@ -0,0 +1,199 @@
+
+	// Create our actions
+#define add_action(varname, actionname, actiontype, toplevel_paths) \
+	Ref<OpenXRAction>##varname;                                     \
```
Seems to fail because of missing space?

```suggestion
	Ref<OpenXRAction> ##varname;                                    \
```

BTW we usually name our macros in uppercase.

BastiaanOlij replied:
That space was removed by the clang-format I think... :(

akien-mga replied:
Indeed, what a mess. That should work (also moved it in scope because I didn't see a reason for it to be out?):
```diff
diff --git a/modules/openxr/openxr_action_sets.cpp b/modules/openxr/openxr_action_sets.cpp
index 9d6e23ae56..867f5547cf 100644
--- a/modules/openxr/openxr_action_sets.cpp
+++ b/modules/openxr/openxr_action_sets.cpp
@@ -152,9 +152,10 @@ void OpenXRActionSets::create_default_action_sets() {
 	add_action_set(action_set);
 
 	// Create our actions
+	// clang-format off
 #define add_action(varname, actionname, actiontype, toplevel_paths) \
-	Ref<OpenXRAction>##varname;                                     \
 	{                                                               \
+		Ref<OpenXRAction> ##varname;                                \
 		##varname.instantiate();                                    \
 		##varname->set_name(String(#varname));                      \
 		##varname->set_localised_name(String(actionname));          \
@@ -162,6 +163,7 @@ void OpenXRActionSets::create_default_action_sets() {
 		##varname->parse_toplevel_paths(String(toplevel_paths));    \
 		action_set->add_action(##varname);                          \
 	}
+	// clang-format on
 
 	add_action(trigger, "Trigger", OpenXRAction::OPENXR_ACTION_FLOAT, "/user/hand/left,/user/hand/right");
 	add_action(trigger_click, "Trigger Click", OpenXRAction::OPENXR_ACTION_BOOL, "/user/hand/left,/user/hand/right");
```

BastiaanOlij replied:
@akien-mga I checked it in with the space but clang format is indeed complaining about it on my computer. We'll see what ci has to say about it

akien-mga replied:
See my diff again, I'm specifically disabling clang-format with `// clang-format off` (and re-enabling with `// clang-format on`).

BastiaanOlij commented:
@akien-mga the variable definitions are "out of scope" on purpose, because I will be using them later on ;) code isn't finished yet

BastiaanOlij commented:
Why do linux compilers always give me trouble... geez :)

BastiaanOlij commented on modules/openxr/doc_classes/OpenXRInterface.xml:
```diff
@@ -0,0 +1,13 @@
+		Due to the needs of OpenXR this interface works slightly different then other plugin based XR interfaces. It needs to be initialised when Godot starts. You need to enable OpenXR, settings for this can be found in your games project settings under the XR heading. You do need to mark a viewport for use with XR in order for Godot to know which render result should be output to the headset.   
+	</description>
+	<tutorials>
+		<link title="OpenXR documentation">$DOCS_URL/tutorials/vr/openxr/index.html</link>
```
I've added the link to the tutorial here, on the stable docs this points to the Godot 3 documentation, on latest its missing, all the more reason to make sure this page gets written soon :)

BastiaanOlij commented:
Ok, just a quick status update here now that things are starting to move fast.

Rendering now works and it looks like we're tracking properly. Rendering for now still requires the mobile renderer even on desktop (though there is a PR that makes part of the desktop renderer work) and uses a final blit from Godots rendering buffers to OpenXRs swapchain. This does render OpenXRs foveated rendering inoperable, I am looking into a solution for rendering directly into the swapchain but even then OpenXRs approach to foveated rendering does not agree with Godots render setup and I feel only VRS will solve this properly. That is next on my list once this PR is finished.
The other open issue around rendering is OpenXRs insistence that we supply images in linear color space while most devices use 32bit RBGA8 buffers which means banding. Sucks. Right now all output is in sRGB which means the image is over bright inside the headset.
Also note that lighting on Godot primitives is currently broken but that falls outside of the scope of this PR (I am going to look into this however because it has annoyed the heck out of me for some time now).

My current focus is on porting the controller logic over from the plugin. I just committed the changes for the resource classes that allow us to define our OpenXR action map and the code that generates a default map with default interaction profiles. 

Important to note here for those who are not familiar with OpenXR, OpenXR takes ownership over the bindings. Our interaction with OpenXR is over actions. We provide suggestions to OpenXR how we want different controllers to be bound to these actions however whether the XR runtime decides to listen is out of our control. It has big pro's, and big con's, but that fight is long past us and this solution was chosen. The bottom line here is that you have no control, and what you submit to OpenXR at startup can't be altered. Re-binding is handled by the XR runtime (if you are lucky, most currently don't offer this so if your player is left handed, sucks for them).

Next thing I'll be working on is making the action map actually work so we have working controllers, and after that I need to port the UI I was working on. 

Lastly, right now Android is the only platform that isn't working yet. The files are there but I'll need help from someone who knows Android more to put it all together. For Meta Quest support it is also important to find out of the official Khronos loader will be supported by the Quest.

BastiaanOlij commented:
Action map code is mostly complete and it registers the action map with OpenXR. There are still issues with incorrect paths so need to find which ones are wrong.
 
I've got basic tracking logic going and it is now tracking the two controllers.

BastiaanOlij commented:
I've marked this PR ready for review, there is more to do but the core is there and working and I think its better to do the rest in follow up PRs.

BastiaanOlij commented:
Updated to OpenXR 1.0.22 and added the compiler define `DISABLE_STD_FILESYSTEM` when compiling the OpenXR thirdparty files as this was causing issues for some people.

BastiaanOlij commented:
@akien-mga the static checks are failing on the 3rd party files, I thought those were being excluded? Is there something i need to add here?

m4gr3d commented on doc/classes/ProjectSettings.xml:
```diff
@@ -1909,5 +1909,14 @@
+		<member name="xr/openxr/enabled" type="bool" setter="" getter="" default="false">
+			If [code]true[/code] Godot will setup and initialise OpenXR on startup.
+		</member>
+		<member name="xr/openxr/in_editor" type="bool" setter="" getter="" default="false">
+			If [code]true[/code] Godot will setup and initialise OpenXR when Godot is run in editor mode enabling editor VR features (experimental).
+		</member>
```
Exciting! I'm guessing that's just a stub for now?

BastiaanOlij replied:
Yup, just a stub and I might comment it out if we end up merging the PR as is, but there are various places in the code where I've made preparations.

m4gr3d commented on drivers/openxr/SCsub:
```diff
@@ -0,0 +1,108 @@
+
+    # may need to include java parts of the openxr loader
+
+    # temporarily disabled until we finish this logic
+    is_supported_on_platform = False
```
@BastiaanOlij Can you create an issue and assign it to me for the Android portion.

BastiaanOlij replied:
Shall do :)

BastiaanOlij replied:
https://github.com/godotengine/godot/issues/57498

akien-mga commented:
> @akien-mga the static checks are failing on the 3rd party files, I thought those were being excluded? Is there something i need to add here?

As we discussed privately, this is because of two issues:
- The python checks (`misc/scripts/black_format.sh`) were not properly excluding thirdparty files apparently. As a matter of fact, this was just fixed in #55785 so rebasing should fix it.
- Those python files are actually the test suite and tools of the jsoncpp lib, and we don't need them, so they should be removed. But that's not urgent and something I can have a look at myself when I review the thirdparty lib integration (but feel free to do a pass at stripping down the files we don't need and documenting the update workflow).

m4gr3d commented on drivers/openxr/openxr_device.h:
```diff
@@ -0,0 +1,299 @@
+#ifndef OPENXR_DRIVER_H
+#define OPENXR_DRIVER_H
+
+#ifdef OPENXR_DUMMY
+// if OpenXR is not supported on the platform, we still have other parts of Godot
+// attempting to use the OpenXR driver, so we need a dummy
+
+class OpenXRDevice {
+public:
+	static void setup_global_defs(){};
+	static bool openxr_is_enabled() { return false; };
+	static OpenXRDevice *get_singleton() { return nullptr; };
+
+	bool is_initialized() { return false; };
+	bool is_running() { return false; };
+	bool initialise(const String &p_rendering_driver) { return false; };
+	bool initialise_session() { return false; };
+	void finish(){};
+};
+
+#else
```
How complex would it be to turn `OpenXRDevice` into an interface and have a `dummy` and a `concrete` implementation similar to the pattern used in other areas?
That'll ensure the dummy version stays up-to-date as this class grows in functionality.

BastiaanOlij replied:
My main reason for not doing this is that the implementation is purely based on OpenXR, there really isn't a way to abstract things to some sort of template class that makes a lot of sense and especially for the vulkan extension, we'd be implementing far more on the dummy class while it is never even instantiated (the whole point is that if OpenXR is not available or not running, there will never be an instance).

This really is uniquely OpenXR and frankly frankenstein enough, while I understand why OpenXR went down this path, I really don't think they've architectures this very nicely. From a principled view they way OpenXR sets its hooks into Vulkan only makes sense for a purpose build XR engine.

m4gr3d commented on drivers/openxr/openxr_device.h:
```diff
@@ -0,0 +1,299 @@
+#include "openxr_extension_wrapper.h"
+
+// Note, OpenXR code that we wrote for our plugin makes use of C++20 notation for initialising structs which ensures zeroing out unspecified members.
+// Godot is currently restricted to C++17 which doesn't allow this notation. Make sure critical fields are set.
```
@BastiaanOlij @akien-mga Is it possible to leverage `clang-format` to ensure this is properly observed?

akien-mga replied:
`clang-format` is only a formatting tool, it can't change code.

`clang-tidy` can do code analysis and changes, including giving a default initialization to uninitialized variables, though it can be somewhat unwieldy to use (applies to the whole case, has false positives and creates compilation bugs).

But what's this C++20 notation that we can't use here? If it's `{}`, AFAIK it's available in C++17, we use it in some places.

BastiaanOlij replied:
@akien-mga this does not work on C++17, atleast for MSVC:
```
	XrEventDataBuffer runtimeEvent = {
		.type = XR_TYPE_EVENT_DATA_BUFFER,
		.next = nullptr
	};
```

And even on C++20 MSVC has small syntax differences compared to LLVM and MSVC requires you to keep the fields in the order of the struct, while on other compilers you can reorder them. On both though you can leave fields out and they will be zero filled.

That said, this syntax is far more readable and far easier to understand than the C++17 syntax I'm currently using (though the comments help in accomplishing the same :))

akien-mga replied:
I think in C++17 you can do:
```
XrEventDataBuffer runtimeEvent = {};
// or XrEventDataBuffer runtimeEvent {};
runtimeEvent.type = XR_TYPE_EVENT_DATA_BUFFER;
```
The init with `{}` should zero everything in the struct and then you can set the members you care about outside the init.

akien-mga replied:
Tested, works as early as C++11:
```cpp
#include <iostream>

class SomeClass {
public:
	int class_a = 5;
	SomeClass() {}
};

struct SomeStruct {
	int some_int;
	float some_float;
	SomeClass some_obj;
	SomeClass *some_ptr;
};

int main() {
	SomeStruct some_str {};
	std::cout << some_str.some_int << " " << some_str.some_float << " " << some_str.some_obj.class_a << " " << some_str.some_ptr << std::endl;
	SomeStruct other_str;
	std::cout << other_str.some_int << " " << other_str.some_float << " " << other_str.some_obj.class_a << " " << other_str.some_ptr << std::endl;
	return 0;
}
```
```
$ g++ -std=c++11 test.cpp
$ ./a.out
0 0 5 0
-327758120 4.55576e-41 5 0x7effec76d160
```

BastiaanOlij replied:
> The init with `{}` should zero everything in the struct and then you can set the members you care about outside the init.
We found out the hard way (note Godot crashing hard), that this is not true on every platform. Android did not zero out things when initialized this way and it ended up crashing stuff when the next pointer had a random value instead of being zero.

m4gr3d commented on drivers/openxr/openxr_device.cpp:
```diff
@@ -0,0 +1,2386 @@
+	version += String::num_int64(XR_VERSION_PATCH(p_version));
+
+	return version;
+}
```
It may be useful to move the above functions into a util header file to keep this class smaller and concise.

BastiaanOlij replied:
I've moved those not dependent on member variables into a new `openxr_util.h` header file. Does indeed help a bit.

m4gr3d commented on drivers/openxr/openxr_device.cpp:
```diff
@@ -0,0 +1,2386 @@
+		"", // applicationName, we'll set this down below
+		1, // applicationVersion, we don't currently have this
+		"Godot Game Engine", // engineName
+		4, // engineVersion
```
You should be able to use one of the constants defined in `core/version.h`.

BastiaanOlij replied:
Done, combining major, minor and patch into a single int.

BastiaanOlij replied:
And before this is asked, no I can't use OpenXRs make version as that results in a 64bit int while this member is a 32bit int.

BastiaanOlij commented:
> > @akien-mga the static checks are failing on the 3rd party files, I thought those were being excluded? Is there something i need to add here?
> 
> As we discussed privately, this is because of two issues:
> 
>     * The python checks (`misc/scripts/black_format.sh`) were not properly excluding thirdparty files apparently. As a matter of fact, this was just fixed in [Simplify formatting scripts, add a clang-tidy script, and run clang-tidy #55785](https://github.com/godotengine/godot/pull/55785) so rebasing should fix it.
> 
>     * Those python files are actually the test suite and tools of the jsoncpp lib, and we don't need them, so they should be removed. But that's not urgent and something I can have a look at myself when I review the thirdparty lib integration (but feel free to do a pass at stripping down the files we don't need and documenting the update workflow).

I've removed the test function but it is still falling over a few py files in the root of the jsoncpp lib. I did rebase but its still checking thirdparty files.

akien-mga commented:
> I've removed the test function but it is still falling over a few py files in the root of the jsoncpp lib. I did rebase but its still checking thirdparty files.

So it was actually not fixed by #55785 but broken by it. I'll make a PR to fix it.

rayanmargham commented:
@BastiaanOlij From the discord (linuxmaster) great progress here lets hope it gets merged into master

reduz commented:
I am not sure if I understand why we have drivers/openxr and modules/openxr. Sounds like we should want either one or the other?

fire commented:
`main.cpp` changed causing a merge conflict.

BastiaanOlij commented:
> `main.cpp` changed causing a merge conflict.

This will happen many times yet until this is merged @fire , I'll rebase when I have a moment.

BastiaanOlij commented:
> I am not sure if I understand why we have drivers/openxr and modules/openxr. Sounds like we should want either one or the other?

This was one of the things I wanted to discuss with you @reduz , I'd like the whole thing to be in modules but we can't because part of the solution is interwoven with the Vulkan drivers and has code being executed at startup before the module system is even initialized.

We also can't move what is in modules into drivers, there are classes here that don't belong in drivers and there will be more like classes that will be part of the editor.

So I had to split the solution in two with the driver part having purely the code that interfaces with OpenXR itself, and modules the bits that are facing the end user.

BastiaanOlij commented:
Just did a rebase, nothing more.

BastiaanOlij commented:
@reduz @m4gr3d , after talking this over with Reduz last night I spend today moving everything into modules and changing the way things are mixed in to other parts of the system. 

I'm not done yet because I want to look into merging openxr_device into openxr_interface, I'm not sure about this just yet.
It's already worth having a look at how things are structured now even before this merge so feel free to give feedback.

Btw @akien-mga , any idea how to make the CI happy about
`./drivers/vulkan/vulkan_hooks.h:40:7: error: 'VulkanHooks' has virtual functions but non-virtual destructor [-Werror,-Wnon-virtual-dtor,2]`
This error is given on platforms that do not support OpenXR because the only implementation of this template class is in there. 
I guess I could give each function a default implementation?

BastiaanOlij commented:
After trying out a few things I've decided to keep the split between the interface class and the `OpenXRAPI` (was driver) class which is basically a wrapper over the OpenXR API.

Things got really messy trying to combine it especially around the action map and controller logic. This nicely splits it in logic that faces OpenXR and logic that focuses on the Godot side.

BastiaanOlij commented:
@SaracenOne made a small change making the Vulkan version check a warning for now. Godot checks with the Vulkan driver what version of Vulkan is supported on the host system, we select the highest v1.x.x version of Vulkan that is supported by the host (currently usually 1.2.0 on desktop).

We then have an extra check with the OpenXR runtime as OpenXR will also tell us which version of Vulkan the runtime wishes to use. The weird thing here is that Oculus reports it only supports Vulkan 1.0.0.

We believe we can safely ignore this, 1.2.0 is backwards compatible with 1.0.0 so anything the OpenXR runtime wants to use should work, but we don't want to limit ourselves in using newer Vulkan features (if available) just because the OpenXR runtime doesn't want to use those. 

So far this seems to work fine.

m4gr3d commented on modules/openxr/openxr_api.cpp:
```diff
@@ -0,0 +1,2162 @@
+		wrapper->on_state_ready();
+	}
+
+	// TODO emit signal
```
Is this to be added in a follow-up PR?

BastiaanOlij replied:
Yes, I wanted to get the base line working first, there are a bunch of things we're doing in the plugin that either need to still be ported but are waiting on something (like Meta adopting the Khronos loader) and things that need to change.

But seeing this PR already covers a whole bunch of things, plus that this will allow us to divide some of the work

m4gr3d reviewed (commented):
Looks great!

Added some minor feedback.

BastiaanOlij commented on modules/openxr/action_map/openxr_action.cpp:
```diff
@@ -0,0 +1,91 @@
+#include "openxr_action.h"
+
+void OpenXRAction::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("set_localised_name", "localised_name"), &OpenXRAction::set_localised_name);
+	ClassDB::bind_method(D_METHOD("get_localised_name"), &OpenXRAction::get_localised_name);
+	ADD_PROPERTY(PropertyInfo(Variant::STRING, "localised_name"), "set_localised_name", "get_localised_name");
```
That always gets me.... should be with a z, it's that in the OpenXR API too.. I'm just too European :P

BastiaanOlij commented on modules/openxr/openxr_api.cpp:
```diff
@@ -0,0 +1,2162 @@
+	// obtain info about our system, writing this out completely to make CI on Linux happy..
+	void *next_pointer = nullptr;
+	for (OpenXRExtensionWrapper *wrapper : registered_extension_wrappers) {
+		next_pointer = wrapper->set_system_properties_and_get_next_pointer(next_pointer);
```
That makes sense :)

akien-mga commented:
Getting this warning when compiling on Linux:
```
In file included from thirdparty/openxr/src/loader/loader_platform.hpp:17,
                 from thirdparty/openxr/src/loader/api_layer_interface.hpp:18,
                 from thirdparty/openxr/src/loader/api_layer_interface.cpp:10:
thirdparty/openxr/src/common/platform_utils.hpp: In function 'char* detail::ImplGetSecureEnv(const char*)':
thirdparty/openxr/src/common/platform_utils.hpp:60:55: note: '#pragma message: Warning:  Falling back to non-secure getenv for environmentallookups!  Consider updating to a different libc.'
   60 |     "lookups!  Consider updating to a different libc.")
      |                                                       ^
```

We should either add a `common_config.h` with the proper definition for all platforms, or ensure we set them all in the `SCsub` (it's done for Windows but likely something else is needed for other platforms).

akien-mga commented:
Added a commit to remove all the stuff we don't need and don't care about:
```
 126 files changed, 23 insertions(+), 22185 deletions(-)
```

Should be squashed with the first commit eventually so that we don't add files to remove them and make the repository history needlessly bigger. But I kept it separate for now for review. (And I'm not done reviewing the rest of the changes.)

akien-mga commented:
> Getting this warning when compiling on Linux:
> 
> ```
> In file included from thirdparty/openxr/src/loader/loader_platform.hpp:17,
>                  from thirdparty/openxr/src/loader/api_layer_interface.hpp:18,
>                  from thirdparty/openxr/src/loader/api_layer_interface.cpp:10:
> thirdparty/openxr/src/common/platform_utils.hpp: In function 'char* detail::ImplGetSecureEnv(const char*)':
> thirdparty/openxr/src/common/platform_utils.hpp:60:55: note: '#pragma message: Warning:  Falling back to non-secure getenv for environmentallookups!  Consider updating to a different libc.'
>    60 |     "lookups!  Consider updating to a different libc.")
>       |                                                       ^
> ```
> 
> We should either add a `common_config.h` with the proper definition for all platforms, or ensure we set them all in the `SCsub` (it's done for Windows but likely something else is needed for other platforms).

Fixed, `HAVE_SECURE_GETENV` was set for Windows when it's supposed to be set for **non** Windows (I also excluded UWP to be sure).

akien-mga reviewed (approved):
The thirdparty library integration is good now :+1: 
I removed everything we don't need currently, and fixed the bug with not defining `HAVE_SECURE_GETENV` on Unix (https://github.com/godotengine/godot/pull/56394#issuecomment-1048657312).

Didn't review the Vulkan context changes and OpenXR integration code, but @reduz did, so it should be good to go.

akien-mga commented:
Thanks!
